### PR TITLE
Fix validation of expressions in ORDER BY for DISTINCT queries

### LINKED ISSUE #5660: Improper resolution for fully qualified reference to field from inner query
This query should fail because `t` is not in scope for the outer query:

```sql
SELECT t.a FROM (
    SELECT * FROM (VALUES 1) t(a)
);
```

### PR DESCRIPTION
**Pull request by martint**
Fixes two issues:

1) The derived schema was using fully qualified aliases from the underlying tables.
This resulted in the following query succeeding, when it should fail due to
table t not being visible in the outer query: (#5660)

        SELECT t.a FROM (SELECT t.* FROM (VALUES 1) t(a))

2) Validation of DISTINCT with ORDER BY was only considering syntactic aliases, but it
    should also consider implied aliases derived from the names of the underlying columns. (#5647)

Fixes #5660, #5647

### CODE REVIEW

**Review by phd3 — COMMENTED** [review]
Thanks for the detailed explanation, was very helpful for understanding :) 

In addition to the commit description, the PR also seems to add another independent and necessary fix here correcting the AllColumns related implementation in `StatementAnalyzer#getAliases`. e.g. `SELECT DISTINCT t1.* FROM t1 ORDER BY a` would succeed with the `getAliases` change even if we return qualified fields. So should we split this out into two commits ?  or may be also include that in the commit description?

File: presto-main/src/main/java/io/prestosql/sql/analyzer/StatementAnalyzer.java [hunk 1]
```diff
@@ -2214,9 +2213,22 @@ private Scope computeAndAssignOutputScope(QuerySpecification node, Optional<Scop
 
             for (SelectItem item : node.getSelect().getSelectItems()) {
                 if (item instanceof AllColumns) {
-                    List<Field> itemOutputFields = analysis.getSelectAllResultFields((AllColumns) item);
-                    checkNotNull(itemOutputFields, "output fields is null for select item %s", item);
```
[the remark this answers is not in the document]
Comment (martint) [reply presto-main/src/main/java/io/prestosql/sql/analyzer/StatementAnalyzer.java:2218]: No, that was unintentional.

File: presto-main/src/main/java/io/prestosql/sql/analyzer/StatementAnalyzer.java [hunk 2]
```diff
@@ -2214,9 +2213,22 @@ private Scope computeAndAssignOutputScope(QuerySpecification node, Optional<Scop
...
-                    outputFields.addAll(itemOutputFields);
+                    AllColumns allColumns = (AllColumns) item;
+
+                    List<Field> fields = analysis.getSelectAllResultFields(allColumns);
+                    for (int i = 0; i < fields.size(); i++) {
+                        Field field = fields.get(i);
+
+                        Optional<String> name;
+                        if (!allColumns.getAliases().isEmpty()) {
+                            name = Optional.of(canonicalize(allColumns.getAliases().get(i)));
```
[the remark this answers is not in the document]
Comment (martint) [reply presto-main/src/main/java/io/prestosql/sql/analyzer/StatementAnalyzer.java:2224]: Here we need to extract a name (String) from the AST `Identifier` to assign to the resulting output field. The values returned by getAliases are used for matching purposes. I agree that it looks a bit awkward, but making things more coherent requires a bigger refactor than what this PR is trying to fix.

**Comment by martint** [comment]
>  So should we split this out into two commits ? or may be also include that in the commit description?

You're right. I initially thought I couldn't fix them separately, but it looks like that's doable. I'll split it into two commits.

**Review by phd3 — APPROVED** [review]
Looks good!

**Pull request merged**